In [ ]:
#| default_exp display

# display
> display helpers

In [ ]:
#| export
from pathlib import Path
from typing import Mapping, Sequence
from fastcore.foundation import AttrDict, patch
from fastcore.xml import FT
from fasthtml.components import Div, Details, Summary, Ul, Li, Span, show
from fasthtml.xtend import Style
from dialoghelper.core import add_html
from pote.common import is_listy, flatten, shorten

In [ ]:
import json
from typing import Any
from typing import Literal
from functools import partial
from pathlib import Path
from typing import cast
from IPython.display import HTML

import fastcore.all as FC
from fastcore.foundation import noop, patch
from fastcore.xml import to_xml
from fastcore.test import *

from pote.common import *
from pote.dutil import link_msg

In [ ]:
#| export
_n = '\n'

## ansi color codes
> Helpers and utils to work with ansi codes in notebooks.

What does the above bash script do?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

It's a comprehensive **ANSI escape code generator** for bash terminals. It provides a library of functions (and a CLI) to control terminal output styling and behaviour, including:

- **Text formatting** — bold, faint, italic, underline, blink, strikethrough, inverse, invisible, etc.
- **Foreground & background colours** — standard 8 colours, "intense" variants, 256-colour mode, and full RGB (24-bit)
- **Cursor control** — move up/down/forward/backward, save/restore position, show/hide cursor, jump to specific row/column
- **Display manipulation** — erase lines/display, insert/delete lines or characters, scroll up/down
- **Terminal reporting** — query cursor position, window size (chars and pixels), window state, title, icon name (requires terminal support, reads back from stdin)
- **Miscellaneous** — set window title/icon, ring the bell, print colour tables/code charts, full terminal reset

It can be **sourced** as a library (to use `ansi::*` functions directly in other scripts) or **run directly** as a CLI tool where options like `--bold --red "Hello"` emit the appropriate escape sequences around the text, then automatically restore the terminal state afterward (unless `--no-restore` is passed).

<details class='token-usage-details'><summary>$0.1282</summary>

`total=24,593 | in=24,292 | out=301 | cached=0.0% | cache_new=10,747 | $0.1282`

</details>


Is there something similar directly in Python, stdlib or external package?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Yes, several options:

**stdlib:**
- `curses` — full terminal control (cursor, colors, input), but complex and low-level

**External packages (much more practical):**
- **`rich`** — the most popular; pretty printing, colored text, tables, progress bars, markdown rendering in terminal, syntax highlighting
- **`colorama`** — simple cross-platform ANSI color support (especially useful on Windows)
- **`blessed`** / **`urwid`** — full curses-like terminal control in a friendlier API
- **`prompt_toolkit`** — advanced terminal UIs, used by IPython/Jupyter under the hood

For most use cases, **`rich`** is the go-to choice — it covers the vast majority of what that bash script does (colors, styles, cursor control via its `Console` and `Live` APIs) and is widely used and well-maintained.

<details class='token-usage-details'><summary>$0.0883</summary>

`total=24,904 | in=24,663 | out=241 | cached=43.6% | cache_new=13,720 | $0.0883`

</details>


I know about `rich` and it's pre-installed here in Solveit, but the problem is it converts ansi codes to HTML in Solveit (or general Jupyter Notebooks) so the output is huge. I want pure ansi codes, SolveIt outputs directly support them.

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

For pure ANSI strings with no rendering layer, the lightest options are:

- **`termcolor`** — very simple, just wraps strings with ANSI codes via `colored()`/`cprint()`
- **`sty`** — composable style objects (e.g. `fg.red + bg.blue + ef.bold`), pure strings, no magic
- **Raw codes directly** — since you just need `\033[...m` sequences, a tiny hand-rolled helper is often enough and has zero dependencies

`sty` is the nicest of the packages for this use case since styles are just string prefixes you compose freely. Would you like to go with that, or roll a minimal helper?

<details class='token-usage-details'><summary>$0.1485</summary>

`total=25,356 | in=24,991 | out=365 | cached=0.0% | cache_new=24,746 | $0.1485`

</details>


Ok, for now I just need the table of all ansi colors. Please, replicate in Python the `ansi::colorTable()` function from the bash script above.

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

```python
ESC = '\033['
def sgr(*c): return f"{ESC}{';'.join(map(str,c))}m"

def color_table():
    fgs   = [30,31,32,33,34,35,36,37]
    fgs_i = [90,91,92,93,94,95,96,97]
    bgs = [('(none)',49),('bg-black',40),('+ intense',100),('bg-red',41),('+ intense',101),
           ('bg-green',42),('+ intense',102),('bg-yellow',43),('+ intense',103),
           ('bg-blue',44),('+ intense',104),('bg-magenta',45),('+ intense',105),
           ('bg-cyan',46),('+ intense',106),('bg-white',47),('+ intense',107)]
    def cell(fg, fg_i, bg):
        return (sgr(bg,fg)+sgr(2)+'f'+sgr(22)+'n'+sgr(1)+'b'+sgr(22) +
                sgr(fg_i)+sgr(2)+'F'+sgr(22)+'N'+sgr(1)+'B'+sgr(22)+sgr(0))
    print(f"{'':12}  black    red   green  yellow    blue magenta   cyan  white")
    for lbl, bg in bgs:
        print(f'{lbl:<12}' + ' '.join(cell(fg, fg_i, bg) for fg,fg_i in zip(fgs,fgs_i)))

color_table()
```

<details class='token-usage-details'><summary>$0.0790</summary>

`total=28,149 | in=25,248 | out=2,901 | cached=98.0% | cache_new=283 | $0.0790`

</details>


In [ ]:
#| export
ESC = '\033['
def sgr(*c): return f"{ESC}{';'.join(map(str,c))}m"

def color_table():
    fgs   = [30,31,32,33,34,35,36,37]
    fgs_i = [90,91,92,93,94,95,96,97]
    bgs = [('(none)',49),('bg-black',40),('+ intense',100),('bg-red',41),('+ intense',101),
           ('bg-green',42),('+ intense',102),('bg-yellow',43),('+ intense',103),
           ('bg-blue',44),('+ intense',104),('bg-magenta',45),('+ intense',105),
           ('bg-cyan',46),('+ intense',106),('bg-white',47),('+ intense',107)]
    def cell(fg, fg_i, bg):
        return (sgr(bg,fg)+sgr(2)+'f'+sgr(22)+'n'+sgr(1)+'b'+sgr(22) +
                sgr(fg_i)+sgr(2)+'F'+sgr(22)+'N'+sgr(1)+'B'+sgr(22)+sgr(0))
    print(f"{'':12} black   red  green  yellow  blue  magenta cyan   white")
    for lbl, bg in bgs:
        print(f'{lbl:<12}' + ' '.join(cell(fg, fg_i, bg) for fg,fg_i in zip(fgs,fgs_i)))

In [ ]:
color_table()

             black   red  green  yellow  blue  magenta cyan   white
(none)      fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB
bg-black    fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB
+ intense   fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB
bg-red      fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB
+ intense   fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB
bg-green    fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB
+ intense   fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB
bg-yellow   fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB
+ intense   fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB
bg-blue     fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB
+ intense   fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB
bg-magenta  fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB
+ intense   fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB
bg-cyan     fnbFNB fnbFNB fnbFNB fnbFNB fnbFNB f

Ok, now same with `ansi::colorCodes()` function to get the actual codes.

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

The ANSI codes are being stripped in cell output. Looking at the bash output, each "patch" is just a colored background block showing the code number. Here's the Python version:

```python
def color_codes():
    def patch(code): return f"{sgr(48,5,code)} {code:3} {sgr(49)}"
    
    print('Standard: ' + sgr(1) + sgr(37) + ''.join(patch(c) + (sgr(30) if c==6 else '') for c in range(8)) + sgr(0))
    print('Intense:  ' + sgr(37) + ''.join(patch(c) + (sgr(30) if c==8 else '') for c in range(8,16)) + sgr(0))
    print()
    for i in [16, 22, 28]:
        for j in [i, i+36, i+72, i+108, i+144, i+180]:
            print(sgr(1,37) + ''.join(patch(c) for c in range(j,j+6)) + sgr(0) + '    ' +
                  sgr(30) + ''.join(patch(c) for c in range(j+18,j+24)) + sgr(0))
        print()
    print('Grays:    ' + sgr(1,37) + ''.join(patch(c) for c in range(232,244)) + sgr(0))
    print('          ' + sgr(30)   + ''.join(patch(c) for c in range(244,256)) + sgr(0))

color_codes()
```

<details class='token-usage-details'><summary>$0.1636</summary>

`total=27,549 | in=26,943 | out=606 | cached=0.0% | cache_new=25,522 | $0.1636`

</details>


In [ ]:
#| export
def color_codes():
    def patch(code): return f"{sgr(48,5,code)} {code:3} {sgr(49)}"
    
    print('Standard: ' + sgr(1) + sgr(37) + ''.join(patch(c) + (sgr(30) if c==6 else '') for c in range(8)) + sgr(0))
    print('Intense:  ' + sgr(37) + ''.join(patch(c) + (sgr(30) if c==8 else '') for c in range(8,16)) + sgr(0))
    print()
    for i in [16, 22, 28]:
        for j in [i, i+36, i+72, i+108, i+144, i+180]:
            print(sgr(1,37) + ''.join(patch(c) for c in range(j,j+6)) + sgr(0) + '    ' +
                  sgr(30) + ''.join(patch(c) for c in range(j+18,j+24)) + sgr(0))
        print()
    print('Grays:    ' + sgr(1,37) + ''.join(patch(c) for c in range(232,244)) + sgr(0))
    print('          ' + sgr(30)   + ''.join(patch(c) for c in range(244,256)) + sgr(0))

In [ ]:
color_codes()

Standard:    0    1    2    3    4    5    6    7 
Intense:     8    9   10   11   12   13   14   15 

  16   17   18   19   20   21       34   35   36   37   38   39 
  52   53   54   55   56   57       70   71   72   73   74   75 
  88   89   90   91   92   93      106  107  108  109  110  111 
 124  125  126  127  128  129      142  143  144  145  146  147 
 160  161  162  163  164  165      178  179  180  181  182  183 
 196  197  198  199  200  201      214  215  216  217  218  219 

  22   23   24   25   26   27       40   41   42   43   44   45 
  58   59   60   61   62   63       76   77   78   79   80   81 
  94   95   96   97   98   99      112  113  114  115  116  117 
 130  131  132  133  134  135      148  149  150  151  152  153 
 166  167  168  169  170  171      184  185  186  187  188  189 
 202  203  204  205  206  207      220  221  222  223  224  225 

  28   29   30   31   32   33       46   47   48   49   50   51 
  64   65   66   67   68   69       82   83   84  

In [ ]:
for code in range(256):
    print(f'\033[38;5;{code}m\\033[38;5;{code}m\033[0m')

\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033
\033


In [ ]:
code = 250
print(f"\033[38;5;{code}m" + 'hello world' + '\033[0m')

hello world


In [ ]:
code = 34
print(f"\033[{code}m" + 'hello world' + '\033[0m')

hello world


## folder walker

A function that given a dir path and optional list of exclude names returns a nested collection of paths in this form:
```python
coll = {
    'Apollo astronauts': [
        'Neil Armstrong',
        'Alan Bean',
        {
            'other': [
                'Bruce Wayne',
                'Clark Kent',
                'Peter Parker'
            ]
        },
        'Edgar Mitchell', 
        'Alan Shepard'
    ],
    'Apollo 11': [
        'Neil Armstrong', 
        'Buzz Aldrin', 
        'Edgar Mitchell', 
        {
            'a':
                1
        },
        'Alan Shepard'
    ]
}
```
where folders are a dict containing one key with the folder name and the value is a list of the folder's contents.

In [ ]:
def dir_tree(path: Path, exclude: list[str] | None = None) -> list:
    path = Path(path)
    exclude = set(exclude or [])
    def _collect(p):
        items = sorted(p.iterdir(), key=lambda x: (x.is_file(), x.name))
        res = []
        for o in items:
            if o.name in exclude: continue
            res.append(_collect(o) if o.is_dir() else o)
        return {p.name: res} if p != path else res
    return _collect(path)

In [ ]:
tr = dir_tree(Path('..'), ['.git', '.github', '__pycache__'])
len(tr),tr[2],tr[10]

(11,
 {'pote': [Path('../pote/__init__.py'),
   Path('../pote/_modidx.py'),
   Path('../pote/all.py'),
   Path('../pote/basic.py'),
   Path('../pote/callback.py'),
   Path('../pote/common.py'),
   Path('../pote/config.py'),
   Path('../pote/display.py'),
   Path('../pote/dutil.py'),
   Path('../pote/flakes.py'),
   Path('../pote/gitutil.py'),
   Path('../pote/isolated.py'),
   Path('../pote/logger.py'),
   Path('../pote/logger_loguru.py'),
   Path('../pote/project.py'),
   Path('../pote/server.py'),
   Path('../pote/test.py'),
   Path('../pote/widget.py')]},
 Path('../pyrightconfig.json'))

Write a simple visualizer of `dir_tree` results.

In [ ]:
def _render_tree(o):
    if isinstance(o, Path): return Li(o.name)
    if isinstance(o, dict):
        name,children = next(iter(o.items()))
        return Li(Details(Summary(name), Ul(*[_render_tree(c) for c in children])))

def show_tree(tree): show(Ul(*[_render_tree(o) for o in tree]))

In [ ]:
show_tree(tr)

HTML(<ul>
  <li>
<details><summary>explorer</summary>      <ul>
        <li>
<details><summary>importdlg</summary>            <ul>
              <li>
<details><summary>test_dialogs</summary>                  <ul>
                    <li>TEMPLATE.ipynb</li>
                    <li>complex.ipynb</li>
                    <li>math_funcs.ipynb</li>
                    <li>simple.ipynb</li>
                    <li>symbols.ipynb</li>
                  </ul>
</details>              </li>
              <li>importdlg.ipynb</li>
            </ul>
</details>        </li>
        <li>
<details><summary>isolated</summary>            <ul>
              <li>00_core_isolated.ipynb</li>
              <li>10_isolated_dup1.ipynb</li>
              <li>10_isolated_dup10.ipynb</li>
              <li>10_isolated_dup11.ipynb</li>
              <li>10_isolated_dup2.ipynb</li>
              <li>10_isolated_dup3.ipynb</li>
              <li>10_isolated_dup4.ipynb</li>
              <li>10_isolated_dup5.ipynb</li>
              <li>10_isolated_dup6.ipynb</li>
              <li>10_isolated_dup7.ipynb</li>
              <li>10_isolated_dup8.ipynb</li>
              <li>10_isolated_dup9.ipynb</li>
              <li>isolated.ipynb</li>
              <li>test_isolate_1.ipynb</li>
            </ul>
</details>        </li>
        <li>
<details><summary>js</summary>            <ul>
              <li>attachment_upload.js</li>
              <li>hl_find.js</li>
              <li>jquery-micro-utils.js</li>
              <li>monedit.js</li>
              <li>output.css</li>
              <li>selection.js</li>
            </ul>
</details>        </li>
        <li>
<details><summary>super_w</summary>            <ul>
              <li>super_w.ipynb</li>
              <li>super_w.py</li>
              <li>super_w_dup1.ipynb</li>
              <li>super_w_old.ipynb</li>
            </ul>
</details>        </li>
        <li>.sesskey</li>
        <li>01_htmx_bridge_spinner.ipynb</li>
        <li>01_htmx_bridge_spinner2.ipynb</li>
        <li>01_logger_bck4.ipynb</li>
        <li>dialoghelper_patcher.ipynb</li>
        <li>gcolor.ipynb</li>
        <li>htmx_bridge.ipynb</li>
        <li>linked_msg.ipynb</li>
        <li>observer.ipynb</li>
        <li>registry_pattern.ipynb</li>
        <li>scheduled_run.ipynb</li>
        <li>tag.ipynb</li>
        <li>tirar.ipynb</li>
      </ul>
</details>  </li>
  <li>
<details><summary>nbs</summary>      <ul>
        <li>
<details><summary>data</summary>            <ul>
              <li>log_across.ipynb</li>
              <li>test.ipynb</li>
            </ul>
</details>        </li>
        <li>
<details><summary>static</summary>            <ul>
              <li>apollo_astronauts.json</li>
              <li>file.txt</li>
              <li>wordlist.txt</li>
            </ul>
</details>        </li>
        <li>00_basic.ipynb</li>
        <li>00_dutil.ipynb</li>
        <li>00_project.ipynb</li>
        <li>01_display.ipynb</li>
        <li>01_display_corrupted.ipynb</li>
        <li>01_flakes.ipynb</li>
        <li>02_git.ipynb</li>
        <li>02_isolated.ipynb</li>
        <li>02_logger.ipynb</li>
        <li>02_logger_loguru.ipynb</li>
        <li>02_server.ipynb</li>
        <li>05_test.ipynb</li>
        <li>10_callback.ipynb</li>
        <li>15_config.ipynb</li>
        <li>17_display.ipynb</li>
        <li>20_widgets.ipynb</li>
        <li>_quarto.yml</li>
        <li>apilist.txt</li>
        <li>index.ipynb</li>
        <li>llms-ctx-full.txt</li>
        <li>llms-ctx.txt</li>
        <li>llms.txt</li>
        <li>nbdev.yml</li>
        <li>styles.css</li>
      </ul>
</details>  </li>
  <li>
<details><summary>pote</summary>      <ul>
        <li>__init__.py</li>
        <li>_modidx.py</li>
        <li>all.py</li>
        <li>basic.py</li>
        <li>callback.py</li>
        <li>common.py</li>
        <li>config.py</li>
        <li>display.py</li>
        <li>dutil.py</li>
        <li>flakes.py</li>
        <li>gitutil.py</li>

Why have you decided to use FastHTML` and `Details` elements instead of simple prints?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

A few reasons:

- **Collapsible folders** — `Details`/`Summary` gives you interactive expand/collapse for free, which is really valuable for deep or wide trees
- **Rendered in the output cell** — solveit renders HTML output inline, so you get a much nicer visual result than text
- **No truncation** — plain `print` in notebooks often gets clipped for large outputs, whereas HTML output scrolls cleanly
- **Scalable** — your tree had quite a few files; a text-based print would have been a wall of text

That said, if you'd prefer a simpler text-based approach (e.g. with indentation and `├──` style characters), that's easy to do too.

<details class='token-usage-details'><summary>$0.0319</summary>

`total=17,522 | in=17,317 | out=205 | cached=75.5% | cache_new=2,013 | $0.0319`

</details>


No, I like it, but give me also the text-based approach (with optional color ansi codes).

In [ ]:
def print_tree(tree, indent=0, color=True):
    _dir = '\033[34m' if color else ''
    _file = '\033[0m' if color else ''
    _reset = '\033[0m' if color else ''
    for o in tree:
        if isinstance(o, Path): print(' ' * indent + _file + o.name + _reset)
        else:
            name,children = next(iter(o.items()))
            print(' ' * indent + _dir + name + _reset)
            print_tree(children, indent+2, color)

In [ ]:
print_tree(tr[0]['explorer'])

importdlg
  test_dialogs
    TEMPLATE.ipynb
    complex.ipynb
    math_funcs.ipynb
    simple.ipynb
    symbols.ipynb
  importdlg.ipynb
isolated
  00_core_isolated.ipynb
  10_isolated_dup1.ipynb
  10_isolated_dup10.ipynb
  10_isolated_dup11.ipynb
  10_isolated_dup2.ipynb
  10_isolated_dup3.ipynb
  10_isolated_dup4.ipynb
  10_isolated_dup5.ipynb
  10_isolated_dup6.ipynb
  10_isolated_dup7.ipynb
  10_isolated_dup8.ipynb
  10_isolated_dup9.ipynb
  isolated.ipynb
  test_isolate_1.ipynb
js
  attachment_upload.js
  hl_find.js
  jquery-micro-utils.js
  monedit.js
  output.css
  selection.js
super_w
  super_w.ipynb
  super_w.py
  super_w_dup1.ipynb
  super_w_old.ipynb
.sesskey
01_htmx_bridge_spinner.ipynb
01_htmx_bridge_spinner2.ipynb
01_logger_bck4.ipynb
dialoghelper_patcher.ipynb
gcolor.ipynb
htmx_bridge.ipynb
linked_msg.ipynb
observer.ipynb
registry_pattern.ipynb
scheduled_run.ipynb
tag.ipynb
tirar.ipynb


what about the `├──` style characters?

In [ ]:
def print_tree(tree, prefix='', color=True):
    _dir = '\033[34m' if color else ''
    _reset = '\033[0m' if color else ''
    for i,o in enumerate(tree):
        last = i == len(tree)-1
        connector = '└── ' if last else '├── '
        extension = '    ' if last else '│   '
        if isinstance(o, Path): print(prefix + connector + o.name)
        else:
            name,children = next(iter(o.items()))
            print(prefix + connector + _dir + name + _reset)
            print_tree(children, prefix+extension, color)

In [ ]:
print_tree(tr[1]['nbs'])

├── data
│   ├── log_across.ipynb
│   └── test.ipynb
├── static
│   ├── apollo_astronauts.json
│   ├── file.txt
│   └── wordlist.txt
├── 00_basic.ipynb
├── 00_dutil.ipynb
├── 00_project.ipynb
├── 01_display.ipynb
├── 01_display_corrupted.ipynb
├── 01_flakes.ipynb
├── 02_git.ipynb
├── 02_isolated.ipynb
├── 02_logger.ipynb
├── 02_logger_loguru.ipynb
├── 02_server.ipynb
├── 05_test.ipynb
├── 10_callback.ipynb
├── 15_config.ipynb
├── 17_display.ipynb
├── 20_widgets.ipynb
├── _quarto.yml
├── apilist.txt
├── index.ipynb
├── llms-ctx-full.txt
├── llms-ctx.txt
├── llms.txt
├── nbdev.yml
└── styles.css


`dir_tree` sort results in lexicographical order, right?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Yes, with dirs before files (due to the `(x.is_file(), x.name)` sort key), then lexicographically within each group.

<details class='token-usage-details'><summary>$0.1167</summary>

`total=20,836 | in=20,746 | out=90 | cached=0.0% | cache_new=18,905 | $0.1167`

</details>


Hmmm, I'd prefer something more like macOS, lexicographical order irrespective of file/folder.

In [ ]:
def dir_tree(path: Path, exclude: list[str] | None = None) -> list:
    path = Path(path)
    exclude = set(exclude or [])
    def _collect(p):
        res = []
        for o in sorted(p.iterdir(), key=lambda x: x.name.lower()):
            if o.name in exclude: continue
            res.append(_collect(o) if o.is_dir() else o)
        return {p.name: res} if p != path else res
    return _collect(path)

Let's see a generator version.

In [ ]:
def dir_tree(path, exclude: list[str] | None = None) -> list[dict|Path]:
    path = Path(path)
    exclude = set(exclude or [])
    def _collect(p):
        for o in sorted(p.iterdir(), key=lambda x: x.name.lower()):
            if o.name in exclude: continue
            yield {o.name: list(_collect(o))} if o.is_dir() else o
    return _collect(path)

In [ ]:
tr = list(dir_tree('..', exclude=['.git', '__pycache__', '.ipynb_checkpoints']))
len(tr), tr[5]

(12,
 {'nbs': [Path('../nbs/00_basic.ipynb'),
   Path('../nbs/00_dutil.ipynb'),
   Path('../nbs/00_project.ipynb'),
   Path('../nbs/01_display.ipynb'),
   Path('../nbs/01_display_corrupted.ipynb'),
   Path('../nbs/01_flakes.ipynb'),
   Path('../nbs/02_git.ipynb'),
   Path('../nbs/02_isolated.ipynb'),
   Path('../nbs/02_logger.ipynb'),
   Path('../nbs/02_logger_loguru.ipynb'),
   Path('../nbs/02_server.ipynb'),
   Path('../nbs/05_test.ipynb'),
   Path('../nbs/10_callback.ipynb'),
   Path('../nbs/15_config.ipynb'),
   Path('../nbs/17_display.ipynb'),
   Path('../nbs/20_widgets.ipynb'),
   Path('../nbs/_quarto.yml'),
   Path('../nbs/apilist.txt'),
   {'data': [Path('../nbs/data/log_across.ipynb'),
     Path('../nbs/data/test.ipynb')]},
   Path('../nbs/index.ipynb'),
   Path('../nbs/llms-ctx-full.txt'),
   Path('../nbs/llms-ctx.txt'),
   Path('../nbs/llms.txt'),
   Path('../nbs/nbdev.yml'),
   {'static': [Path('../nbs/static/apollo_astronauts.json'),
     Path('../nbs/static/file.txt'),
  

Why haven't you used `Path.walk` to create the tree? It's just curiosity.

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

`Path.walk` yields `(dirpath, dirnames, filenames)` tuples in a flat, top-down traversal. Building a **nested** structure from it is awkward — you'd need to maintain a mapping of paths to their parent lists to know where to insert each node. The recursive `_collect` approach naturally mirrors the nested structure we want, so it's simpler and cleaner here.

<details class='token-usage-details'><summary>$0.1301</summary>

`total=23,276 | in=23,185 | out=91 | cached=0.0% | cache_new=21,060 | $0.1301`

</details>


Fair enough, but now I'd like for `tree_dir` to call optional callbacks whenever it founds a folder or a file.

In [ ]:
def dir_tree(path, exclude: list[str] | None = None, on_file=None, on_dir=None):
    path = Path(path)
    exclude = set(exclude or [])
    def _collect(p):
        for o in sorted(p.iterdir(), key=lambda x: x.name.lower()):
            if o.name in exclude: continue
            if o.is_dir():
                if on_dir: on_dir(o)
                yield {o.name: list(_collect(o))}
            else:
                yield o if not on_file else on_file(o)
    return _collect(path)

In [ ]:
tr = list(dir_tree('..', exclude=['.git', '__pycache__', '.ipynb_checkpoints'], on_file=lambda o: o.name))
tr[5]

{'nbs': ['00_basic.ipynb',
  '00_dutil.ipynb',
  '00_project.ipynb',
  '01_display.ipynb',
  '01_display_corrupted.ipynb',
  '01_flakes.ipynb',
  '02_git.ipynb',
  '02_isolated.ipynb',
  '02_logger.ipynb',
  '02_logger_loguru.ipynb',
  '02_server.ipynb',
  '05_test.ipynb',
  '10_callback.ipynb',
  '15_config.ipynb',
  '17_display.ipynb',
  '20_widgets.ipynb',
  '_quarto.yml',
  'apilist.txt',
  {'data': ['log_across.ipynb', 'test.ipynb']},
  'index.ipynb',
  'llms-ctx-full.txt',
  'llms-ctx.txt',
  'llms.txt',
  'nbdev.yml',
  {'static': ['apollo_astronauts.json', 'file.txt', 'wordlist.txt']},
  'styles.css']}

Hmmm, `dir_tree` callbacks have two functions: side-effects (e.g. print) **and** transformation (e.g. `on_file=lambda o: o.name`). With files that works fine, with folders we only got side-effects, because in fact, `dir_tree` doesn't return folders items, only their names as keys. How would you fix that?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

The fix is to pass both the folder path and its children to `on_dir`, and use its return value as the yielded item (falling back to the default dict if no callback):

```python
def dir_tree(path, exclude=None, on_file=None, on_dir=None):
    path = Path(path)
    exclude = set(exclude or [])
    def _collect(p):
        for o in sorted(p.iterdir(), key=lambda x: x.name.lower()):
            if o.name in exclude: continue
            if o.is_dir():
                children = list(_collect(o))
                yield on_dir(o, children) if on_dir else {o.name: children}
            else:
                yield on_file(o) if on_file else o
    return _collect(path)
```

Now `on_dir` gets `(path, children)` and can both side-effect and transform, e.g. `on_dir=lambda o, ch: {o.name.upper(): ch}`.

<details class='token-usage-details'><summary>$0.1126</summary>

`total=18,491 | in=17,931 | out=560 | cached=0.0% | cache_new=17,121 | $0.1126`

</details>


Yes, that's better, but the problem of not having items that are folders remains. Mayve we should take a lesson from Unix and return '.' items signaling current folder. So, the `is_dir` path the function returns a children list whose first item is the dir itself. What do you think? Let's discuss, don't code just yet.

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

It's a reasonable idea, and Unix's `.` convention is familiar. But I see a few trade-offs:

**In favour:**
- Gives `on_dir` access to the folder path in a uniform way, without needing a separate callback signature
- Makes the structure self-describing — each subtree carries its own root

**Against:**
- It breaks the clean symmetry of the current format: right now every item is either a `Path` (leaf) or a `dict` (subtree), which makes recursive processing straightforward. Adding a sentinel `.` item as a first child is a bit of an implicit convention that callers need to know about
- The current `on_dir(o, children)` signature already solves the problem cleanly — the folder path *is* available to the callback, it just isn't in the output unless the callback puts it there

**Alternative worth considering:** instead of a sentinel, keep the current structure but change the dict to hold both the path and the children — e.g. `{o: children}` with a `Path` key instead of a string. That way the folder path is always present in the output, traversal code still has a clear `dict` vs `Path` distinction, and no sentinel magic is needed.

What's your main motivation — is it about having the full path available post-hoc when traversing the result, or more about the callback interface?

<details class='token-usage-details'><summary>$0.1096</summary>

`total=18,618 | in=18,310 | out=308 | cached=0.0% | cache_new=18,053 | $0.1096`

</details>


That was my original idea, use the dir Path as a key, but Path is not hashable, or is it?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

`Path` actually **is** hashable in Python:

```python
p = Path('foo/bar.txt')
{p: 'some metadata'}
```

<details class='token-usage-details'><summary>$0.0781</summary>

`total=32,357 | in=32,317 | out=40 | cached=55.9% | cache_new=4,379 | $0.0781`

</details>


Great, then that's a much better solution than adding a the sentinel.

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Agreed! Let's update `dir_tree` to use the `Path` as the dict key:

Now traversal code can use `isinstance(o, dict)` as before, but the key is a full `Path` instead of a string — so you get the name via `o.name` as usual, plus the full path for free.

In [ ]:
#| export
def dir_tree(path, exclude=None, on_file=None, on_dir=None):
    path = Path(path)
    exclude = set(exclude or [])
    def _collect(p):
        for o in sorted(p.iterdir(), key=lambda x: x.name.lower()):
            if o.name in exclude: continue
            if o.is_dir():
                children = list(_collect(o))
                yield on_dir(o, children) if on_dir else {o: children}
            else:
                yield on_file(o) if on_file else o
    return _collect(path)

In [ ]:
tr = list(dir_tree('..', exclude=['.git', '__pycache__', '.ipynb_checkpoints'], on_file=lambda o: o.name, on_dir=lambda d,c: {d.name: c}))
tr[5]

{'nbs': ['00_basic.ipynb',
  '00_dutil.ipynb',
  '00_project.ipynb',
  '01_display.ipynb',
  '01_display_corrupted.ipynb',
  '01_flakes.ipynb',
  '02_git.ipynb',
  '02_isolated.ipynb',
  '02_logger.ipynb',
  '02_logger_loguru.ipynb',
  '02_server.ipynb',
  '05_test.ipynb',
  '10_callback.ipynb',
  '15_config.ipynb',
  '17_display.ipynb',
  '20_widgets.ipynb',
  '_quarto.yml',
  'apilist.txt',
  {'data': ['log_across.ipynb', 'test.ipynb']},
  'index.ipynb',
  'llms-ctx-full.txt',
  'llms-ctx.txt',
  'llms.txt',
  'nbdev.yml',
  {'static': ['apollo_astronauts.json', 'file.txt', 'wordlist.txt']},
  'styles.css']}

And now, I think we can easily write versions of `render_tree` and `print_tree` using `dir_tree` and callbacks.

In [ ]:
#| export
def show_tree(path, exclude=None):
    def _file(o): return Li(o.name)
    def _dir(o, ch): return Li(Details(Summary(o.name), Ul(*ch)))
    show(Ul(*dir_tree(path, exclude, on_file=_file, on_dir=_dir)))

In [ ]:
def print_tree(path, exclude=None, prefix='', color=True):
    _dir = '\033[34m' if color else ''
    _reset = '\033[0m' if color else ''
    def _file(o): return o
    def _dir_cb(o, ch): return {o: ch}
    def _print(items, prefix=''):
        for i,o in enumerate(items):
            last = i == len(items)-1
            conn = '└── ' if last else '├── '
            ext = '    ' if last else '│   '
            if isinstance(o, dict):
                p,ch = next(iter(o.items()))
                print(prefix + conn + _dir + p.name + _reset)
                _print(ch, prefix+ext)
            else: print(prefix + conn + o.name)
    _print(list(dir_tree(path, exclude, on_file=_file, on_dir=_dir_cb)), prefix)

In [ ]:
show_tree('..', exclude=['.git', '__pycache__', '.ipynb_checkpoints'])

HTML(<ul>
  <li>
<details><summary>.github</summary>      <ul>
        <li>
<details><summary>workflows</summary>            <ul>
              <li>deploy.yaml</li>
              <li>test.yaml</li>
            </ul>
</details>        </li>
      </ul>
</details>  </li>
  <li>.gitignore</li>
  <li>
<details><summary>explorer</summary>      <ul>
        <li>.sesskey</li>
        <li>01_htmx_bridge_spinner.ipynb</li>
        <li>01_htmx_bridge_spinner2.ipynb</li>
        <li>01_logger_bck4.ipynb</li>
        <li>dialoghelper_patcher.ipynb</li>
        <li>gcolor.ipynb</li>
        <li>htmx_bridge.ipynb</li>
        <li>
<details><summary>importdlg</summary>            <ul>
              <li>importdlg.ipynb</li>
              <li>
<details><summary>test_dialogs</summary>                  <ul>
                    <li>complex.ipynb</li>
                    <li>math_funcs.ipynb</li>
                    <li>simple.ipynb</li>
                    <li>symbols.ipynb</li>
                    <li>TEMPLATE.ipynb</li>
                  </ul>
</details>              </li>
            </ul>
</details>        </li>
        <li>
<details><summary>isolated</summary>            <ul>
              <li>00_core_isolated.ipynb</li>
              <li>10_isolated_dup1.ipynb</li>
              <li>10_isolated_dup10.ipynb</li>
              <li>10_isolated_dup11.ipynb</li>
              <li>10_isolated_dup2.ipynb</li>
              <li>10_isolated_dup3.ipynb</li>
              <li>10_isolated_dup4.ipynb</li>
              <li>10_isolated_dup5.ipynb</li>
              <li>10_isolated_dup6.ipynb</li>
              <li>10_isolated_dup7.ipynb</li>
              <li>10_isolated_dup8.ipynb</li>
              <li>10_isolated_dup9.ipynb</li>
              <li>isolated.ipynb</li>
              <li>test_isolate_1.ipynb</li>
            </ul>
</details>        </li>
        <li>
<details><summary>js</summary>            <ul>
              <li>attachment_upload.js</li>
              <li>hl_find.js</li>
              <li>jquery-micro-utils.js</li>
              <li>monedit.js</li>
              <li>output.css</li>
              <li>selection.js</li>
            </ul>
</details>        </li>
        <li>linked_msg.ipynb</li>
        <li>observer.ipynb</li>
        <li>registry_pattern.ipynb</li>
        <li>scheduled_run.ipynb</li>
        <li>
<details><summary>super_w</summary>            <ul>
              <li>super_w.ipynb</li>
              <li>super_w.py</li>
              <li>super_w_dup1.ipynb</li>
              <li>super_w_old.ipynb</li>
            </ul>
</details>        </li>
        <li>tag.ipynb</li>
        <li>tirar.ipynb</li>
      </ul>
</details>  </li>
  <li>LICENSE</li>
  <li>MANIFEST.in</li>
  <li>
<details><summary>nbs</summary>      <ul>
        <li>00_basic.ipynb</li>
        <li>00_dutil.ipynb</li>
        <li>00_project.ipynb</li>
        <li>01_display.ipynb</li>
        <li>01_display_corrupted.ipynb</li>
        <li>01_flakes.ipynb</li>
        <li>02_git.ipynb</li>
        <li>02_isolated.ipynb</li>
        <li>02_logger.ipynb</li>
        <li>02_logger_loguru.ipynb</li>
        <li>02_server.ipynb</li>
        <li>05_test.ipynb</li>
        <li>10_callback.ipynb</li>
        <li>15_config.ipynb</li>
        <li>17_display.ipynb</li>
        <li>20_widgets.ipynb</li>
        <li>_quarto.yml</li>
        <li>apilist.txt</li>
        <li>
<details><summary>data</summary>            <ul>
              <li>log_across.ipynb</li>
              <li>test.ipynb</li>
            </ul>
</details>        </li>
        <li>index.ipynb</li>
        <li>llms-ctx-full.txt</li>
        <li>llms-ctx.txt</li>
        <li>llms.txt</li>
        <li>nbdev.yml</li>
        <li>
<details><summary>static</summary>            <ul>
              <li>apollo_astronauts.json</li>
              <li>file.txt</li>
              <li>wordlist.txt</li>
            </ul>
</details>        </li>
        <li>styles.css</li>
      </ul>
</details>  </li>
  <li>
<details><summary

In [ ]:
print_tree('.', exclude=['.git', '__pycache__', '.ipynb_checkpoints'])

├── 00_basic.ipynb
├── 00_dutil.ipynb
├── 00_project.ipynb
├── 01_display.ipynb
├── 01_display_corrupted.ipynb
├── 01_flakes.ipynb
├── 02_git.ipynb
├── 02_isolated.ipynb
├── 02_logger.ipynb
├── 02_logger_loguru.ipynb
├── 02_server.ipynb
├── 05_test.ipynb
├── 10_callback.ipynb
├── 15_config.ipynb
├── 17_display.ipynb
├── 20_widgets.ipynb
├── _quarto.yml
├── apilist.txt
├── data
│   ├── log_across.ipynb
│   └── test.ipynb
├── index.ipynb
├── llms-ctx-full.txt
├── llms-ctx.txt
├── llms.txt
├── nbdev.yml
├── static
│   ├── apollo_astronauts.json
│   ├── file.txt
│   └── wordlist.txt
└── styles.css


###### start-section

And now, I think we can easily write versions of `render_tree` and `print_tree` using `dir_tree` and callbacks.

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

The key insight: use `on_file=lambda o: o.name` to pre-transform files, so the render functions deal with plain strings instead of `Path` objects, simplifying the type checks:

In [ ]:
def show_tree(path, exclude=None):
    tr = dir_tree(path, exclude, on_file=lambda o: o.name)
    def _render(o):
        if isinstance(o, str): return Li(o)
        name,children = next(iter(o.items()))
        return Li(Details(Summary(name), Ul(*[_render(c) for c in children])))
    show(Ul(*[_render(o) for o in tr]))

def print_tree(path, exclude=None, color=True):
    tr = dir_tree(path, exclude, on_file=lambda o: o.name)
    _d,_r = ('\033[34m','\033[0m') if color else ('','')
    def _print(items, prefix=''):
        items = list(items)
        for i,o in enumerate(items):
            last = i == len(items)-1
            conn,ext = ('└── ','    ') if last else ('├── ','│   ')
            if isinstance(o, str): print(prefix + conn + o)
            else:
                name,children = next(iter(o.items()))
                print(prefix + conn + _d + name + _r)
                _print(children, prefix+ext)
    _print(tr)

In [ ]:
show_tree('..', exclude=['.git', '__pycache__', '.ipynb_checkpoints'])

HTML(<ul>
  <li>
<details><summary>.github</summary>      <ul>
        <li>
<details><summary>workflows</summary>            <ul>
              <li>deploy.yaml</li>
              <li>test.yaml</li>
            </ul>
</details>        </li>
      </ul>
</details>  </li>
  <li>.gitignore</li>
  <li>
<details><summary>explorer</summary>      <ul>
        <li>.sesskey</li>
        <li>01_htmx_bridge_spinner.ipynb</li>
        <li>01_htmx_bridge_spinner2.ipynb</li>
        <li>01_logger_bck4.ipynb</li>
        <li>dialoghelper_patcher.ipynb</li>
        <li>gcolor.ipynb</li>
        <li>htmx_bridge.ipynb</li>
        <li>
<details><summary>importdlg</summary>            <ul>
              <li>importdlg.ipynb</li>
              <li>
<details><summary>test_dialogs</summary>                  <ul>
                    <li>complex.ipynb</li>
                    <li>math_funcs.ipynb</li>
                    <li>simple.ipynb</li>
                    <li>symbols.ipynb</li>
                    <li>TEMPLATE.ipynb</li>
                  </ul>
</details>              </li>
            </ul>
</details>        </li>
        <li>
<details><summary>isolated</summary>            <ul>
              <li>00_core_isolated.ipynb</li>
              <li>10_isolated_dup1.ipynb</li>
              <li>10_isolated_dup10.ipynb</li>
              <li>10_isolated_dup11.ipynb</li>
              <li>10_isolated_dup2.ipynb</li>
              <li>10_isolated_dup3.ipynb</li>
              <li>10_isolated_dup4.ipynb</li>
              <li>10_isolated_dup5.ipynb</li>
              <li>10_isolated_dup6.ipynb</li>
              <li>10_isolated_dup7.ipynb</li>
              <li>10_isolated_dup8.ipynb</li>
              <li>10_isolated_dup9.ipynb</li>
              <li>isolated.ipynb</li>
              <li>test_isolate_1.ipynb</li>
            </ul>
</details>        </li>
        <li>
<details><summary>js</summary>            <ul>
              <li>attachment_upload.js</li>
              <li>hl_find.js</li>
              <li>jquery-micro-utils.js</li>
              <li>monedit.js</li>
              <li>output.css</li>
              <li>selection.js</li>
            </ul>
</details>        </li>
        <li>linked_msg.ipynb</li>
        <li>observer.ipynb</li>
        <li>registry_pattern.ipynb</li>
        <li>scheduled_run.ipynb</li>
        <li>
<details><summary>super_w</summary>            <ul>
              <li>super_w.ipynb</li>
              <li>super_w.py</li>
              <li>super_w_dup1.ipynb</li>
              <li>super_w_old.ipynb</li>
            </ul>
</details>        </li>
        <li>tag.ipynb</li>
        <li>tirar.ipynb</li>
      </ul>
</details>  </li>
  <li>LICENSE</li>
  <li>MANIFEST.in</li>
  <li>
<details><summary>nbs</summary>      <ul>
        <li>00_basic.ipynb</li>
        <li>00_dutil.ipynb</li>
        <li>00_project.ipynb</li>
        <li>01_display.ipynb</li>
        <li>01_display_corrupted.ipynb</li>
        <li>01_flakes.ipynb</li>
        <li>02_git.ipynb</li>
        <li>02_isolated.ipynb</li>
        <li>02_logger.ipynb</li>
        <li>02_logger_loguru.ipynb</li>
        <li>02_server.ipynb</li>
        <li>05_test.ipynb</li>
        <li>10_callback.ipynb</li>
        <li>15_config.ipynb</li>
        <li>17_display.ipynb</li>
        <li>20_widgets.ipynb</li>
        <li>_quarto.yml</li>
        <li>apilist.txt</li>
        <li>
<details><summary>data</summary>            <ul>
              <li>log_across.ipynb</li>
              <li>test.ipynb</li>
            </ul>
</details>        </li>
        <li>index.ipynb</li>
        <li>llms-ctx-full.txt</li>
        <li>llms-ctx.txt</li>
        <li>llms.txt</li>
        <li>nbdev.yml</li>
        <li>
<details><summary>static</summary>            <ul>
              <li>apollo_astronauts.json</li>
              <li>file.txt</li>
              <li>wordlist.txt</li>
            </ul>
</details>        </li>
        <li>styles.css</li>
      </ul>
</details>  </li>
  <li>
<details><summary

In [ ]:
print_tree('.', exclude=['.git', '__pycache__', '.ipynb_checkpoints'])

├── 00_basic.ipynb
├── 00_dutil.ipynb
├── 00_project.ipynb
├── 01_display.ipynb
├── 01_display_corrupted.ipynb
├── 01_flakes.ipynb
├── 02_git.ipynb
├── 02_isolated.ipynb
├── 02_logger.ipynb
├── 02_logger_loguru.ipynb
├── 02_server.ipynb
├── 05_test.ipynb
├── 10_callback.ipynb
├── 15_config.ipynb
├── 17_display.ipynb
├── 20_widgets.ipynb
├── _quarto.yml
├── apilist.txt
├── data
│   ├── log_across.ipynb
│   └── test.ipynb
├── index.ipynb
├── llms-ctx-full.txt
├── llms-ctx.txt
├── llms.txt
├── nbdev.yml
├── static
│   ├── apollo_astronauts.json
│   ├── file.txt
│   └── wordlist.txt
└── styles.css


###### end-section

In [ ]:
#| export
def print_tree(path, exclude=None, color=True):
    tr = dir_tree(path, exclude, on_file=lambda o: o.name)
    _d,_r,_g = ('\033[34m','\033[0m','\033[38;5;7m') if color else ('','','')
    def _print(items, prefix=''):
        items = list(items)
        for i,o in enumerate(items):
            last = i == len(items)-1
            conn,ext = ('└─ ','   ') if last else ('├─ ','│  ')
            if isinstance(o, str): print(_g + prefix + conn + _r + o)
            else:
                dir,children = next(iter(o.items()))
                print(_g + prefix + conn + _r + _d + dir.name + _r)
                _print(children, prefix+ext)
    _print(tr)

In [ ]:
print_tree('.', exclude=['.git', '__pycache__', '.ipynb_checkpoints'])

├─ 00_basic.ipynb
├─ 00_dutil.ipynb
├─ 00_project.ipynb
├─ 01_display.ipynb
├─ 01_display_corrupted.ipynb
├─ 01_flakes.ipynb
├─ 02_git.ipynb
├─ 02_isolated.ipynb
├─ 02_logger.ipynb
├─ 02_logger_loguru.ipynb
├─ 02_server.ipynb
├─ 05_test.ipynb
├─ 10_callback.ipynb
├─ 15_config.ipynb
├─ 17_display.ipynb
├─ 20_widgets.ipynb
├─ _quarto.yml
├─ apilist.txt
├─ data
│  ├─ log_across.ipynb
│  └─ test.ipynb
├─ index.ipynb
├─ llms-ctx-full.txt
├─ llms-ctx.txt
├─ llms.txt
├─ nbdev.yml
├─ static
│  ├─ apollo_astronauts.json
│  ├─ file.txt
│  └─ wordlist.txt
└─ styles.css


Alright, finally, is there any library out there that does something similar to what we've created here? For example, for `show_tree` we could simply use `tree` command (though much less versatil).

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Yes, a few relevant ones:

- **`tree` command** — as you noted, the simplest option (`!tree -I '__pycache__'`), but no Python integration or customization.

- **[Rich](https://rich.readthedocs.io/en/latest/tree.html)** — probably the closest to our `print_tree`. [*](https://rich.readthedocs.io/en/latest/tree.html "Tree — Rich 14.1.0 documentation") Rich has a `Tree` class that generates a tree view in the terminal, great for presenting filesystem contents or any hierarchical data, where each branch can have a label that's text or any other Rich renderable. It's very polished with colors and styles, but you build the tree manually node by node rather than from a path.

- **[treelib](https://hackersandslackers.com/python-tree-hierachies-treelib/)** — [*](https://hackersandslackers.com/python-tree-hierachies-treelib/ "Visualize Folder Structures with Python’s Treelib") a Python library that allows you to create a visual tree hierarchy, a plaintext representation of parent-child relationships. More general-purpose (any tree, not just filesystem), but again requires manual construction.

- **[directory-tree](https://pypi.org/project/directory-tree/)** — [*](https://pypi.org/project/directory-tree/ "directory-tree · PyPI") a package specifically for printing out the tree structure of a user-defined directory. Simple, but no interactivity or HTML output.

None of them offer the **collapsible HTML output** of our `show_tree`, which is really the most useful feature in a notebook context — that part is genuinely novel.

<details class='token-usage-details'><summary>$0.2727</summary>

`total=94,925 | in=94,446 | out=479 | cached=51.3% | cache_new=29,567 | searches=1 | $0.2727`

</details>


## file metadata

###### start-section

In [ ]:
class FileNode(FC.GetAttr):
    "Node in a file tree"
    _default='path'
    def __init__(self, path, **kwargs):
        self.path = Path(path).resolve(); self._arg = path
        if kwargs: self._meta = kwargs
    def __str__(self): return str(self.path)
    def __repr__(self): return f'{self.__class__.__name__}({self._arg})'
    @property
    def meta(self):
        if not hasattr(self, '_meta'): self._meta = FC.AttrDict()
        return self._meta

In [ ]:
fnd = FileNode('.')
fnd.name, str(fnd), str(fnd.resolve()), type(fnd.resolve())

('nbs', '/app/data/prj/pote/nbs', '/app/data/prj/pote/nbs', pathlib.PosixPath)

In [ ]:
fnd.meta

```python
{}
```

In [ ]:
fnd.meta.open = True
fnd.meta

```python
{'open': True}
```

In [ ]:
tr = list(dir_tree('.', exclude=['.git', '__pycache__', '.ipynb_checkpoints'], on_file=lambda o: FileNode(o)))
tr

[FileNode(00_basic.ipynb),
 FileNode(00_dutil.ipynb),
 FileNode(00_project.ipynb),
 FileNode(01_display.ipynb),
 FileNode(01_display_corrupted.ipynb),
 FileNode(01_flakes.ipynb),
 FileNode(02_git.ipynb),
 FileNode(02_isolated.ipynb),
 FileNode(02_logger.ipynb),
 FileNode(02_logger_loguru.ipynb),
 FileNode(02_server.ipynb),
 FileNode(05_test.ipynb),
 FileNode(10_callback.ipynb),
 FileNode(15_config.ipynb),
 FileNode(17_display.ipynb),
 FileNode(20_widgets.ipynb),
 FileNode(_quarto.yml),
 FileNode(apilist.txt),
 {Path('data'): [FileNode(data/log_across.ipynb), FileNode(data/test.ipynb)]},
 FileNode(index.ipynb),
 FileNode(llms-ctx-full.txt),
 FileNode(llms-ctx.txt),
 FileNode(llms.txt),
 FileNode(nbdev.yml),
 {Path('static'): [FileNode(static/apollo_astronauts.json),
   FileNode(static/file.txt),
   FileNode(static/wordlist.txt)]},
 FileNode(styles.css)]

I want to assign metadata payload to file paths. `FileNode` is a possible solution, but has limitations. How would you implement this?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

The main limitation of `FileNode` is that `isinstance(o, Path)` fails, so it breaks existing code that checks for `Path`. The cleanest fix is to subclass `Path` directly — Python 3.12 supports this without tricks:

###### end-section

In [ ]:
#| export
class FileMeta(Path):
    "Path with attached metadata dict"
    def __new__(cls, *args, **kwargs): return super().__new__(cls, *args)
    def __init__(self, *args, **kwargs):
        super().__init__(*args)
        self.meta = AttrDict(kwargs)

In [ ]:
f = FileMeta('00_basic.ipynb', open=True, tags=['nb'])
isinstance(f, Path), f.name, f.meta

(True, '00_basic.ipynb', {'open': True, 'tags': ['nb']})

In [ ]:
f.__class__.mro()

[__main__.FileMeta, pathlib.Path, pathlib.PurePath, object]

In [ ]:
d = FileMeta('.', open=True, tags=['nb'])
str(d.resolve()), type(d.resolve()), d.is_dir()

('/app/data/prj/pote/nbs', __main__.FileMeta, True)

In [ ]:
tr = list(dir_tree('.', exclude=['.git', '__pycache__', '.ipynb_checkpoints'], on_file=lambda o: FileMeta(o), on_dir=lambda d,c: {FileMeta(d): c}))
tr

[Path('00_basic.ipynb'),
 Path('00_dutil.ipynb'),
 Path('00_project.ipynb'),
 Path('01_display.ipynb'),
 Path('01_display_corrupted.ipynb'),
 Path('01_flakes.ipynb'),
 Path('02_git.ipynb'),
 Path('02_isolated.ipynb'),
 Path('02_logger.ipynb'),
 Path('02_logger_loguru.ipynb'),
 Path('02_server.ipynb'),
 Path('05_test.ipynb'),
 Path('10_callback.ipynb'),
 Path('15_config.ipynb'),
 Path('17_display.ipynb'),
 Path('20_widgets.ipynb'),
 Path('_quarto.yml'),
 Path('apilist.txt'),
 {Path('data'): [Path('data/log_across.ipynb'), Path('data/test.ipynb')]},
 Path('index.ipynb'),
 Path('llms-ctx-full.txt'),
 Path('llms-ctx.txt'),
 Path('llms.txt'),
 Path('nbdev.yml'),
 {Path('static'): [Path('static/apollo_astronauts.json'),
   Path('static/file.txt'),
   Path('static/wordlist.txt')]},
 Path('styles.css')]

`FileMeta` should override some of the methods that return paths to copy over the metadata if necessary. I think these ones:
- absolute, relative_to, resolve
- with_name, with_stem, with_suffix
- copy, copy_into, if `preserve_metadata` is True
- rename
- move, move_into
- replace??, not sure (in fact I don't know what replace does)

But let's discuss.

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

A few thoughts to guide the discussion:

**Always copy metadata (same file, different path representation):**
- `resolve`, `absolute`, `relative_to` — these just change how the same file is addressed, so copying metadata unconditionally makes sense

**Probably don't copy (different file entirely):**
- `with_name`, `with_stem`, `with_suffix` — the result points to a *different* file that may not exist yet or may have nothing to do with the original. Copying metadata silently could be surprising. I'd leave these alone, or only copy if the user explicitly subclasses further.

**`preserve_metadata` flag makes sense:**
- `copy`, `copy_into` — you're duplicating the file, so optionally carrying metadata over is reasonable
- `rename`, `move`, `move_into` — the file is the same content at a new location, so you probably *do* want metadata preserved by default here, unlike copy

**`replace`:** it's like `rename` but overwrites the target atomically if it exists. So it's a destructive move — the *source* file's identity wins. I'd treat it like `rename` and preserve metadata by default.

One broader question worth settling first: should `FileMeta` metadata be considered **file-identity metadata** (travels with the file wherever it goes) or **context metadata** (only meaningful in the specific place it was created)? That determines how aggressive you want to be with propagation.

<details class='token-usage-details'><summary>$0.1513</summary>

`total=25,470 | in=24,687 | out=783 | cached=0.0% | cache_new=20,125 | $0.1513`

</details>


It is context metadata. For instance, a tree visor component can store open/close state or selection state, or allow in-place renaming. The metadata copy-over is a good-to-have helper for that kind of uses, but we can perfectly let the consumer app track the app state itself. So, helpers, not file state.

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

That clarifies things nicely. So `FileMeta` is really about **UI/view state** attached to a path for convenience, not persistent file identity. That simplifies the design considerably:

- **`resolve`, `absolute`, `relative_to`** — worth overriding to preserve metadata, since these are the most common "same file, different representation" operations that a tree viewer would naturally call
- **`with_name`, `with_stem`, `with_suffix`, `replace`** — skip, they imply a different file/identity
- **`rename`, `move`, `move_into`, `copy`, `copy_into`** — skip too; if the app is tracking state itself, it can re-attach metadata after these operations as needed. No need for a `preserve_metadata` flag.

So really just the three path-representation methods are worth overriding. Shall we go ahead with those?

<details class='token-usage-details'><summary>$0.1457</summary>

`total=25,348 | in=25,144 | out=204 | cached=0.0% | cache_new=24,892 | $0.1457`

</details>


In [ ]:
#| export
@patch
def resolve(self:FileMeta, strict=False): return FileMeta(super().resolve(strict=strict), **self.meta)

@patch
def absolute(self:FileMeta): return FileMeta(super().absolute(), **self.meta)

@patch
def relative_to(self:FileMeta, *other): return FileMeta(super().relative_to(*other), **self.meta)

## DetailsColl
> Collapsible dict viewer component for notebooks

A FastHTML component that renders nested dictionaries/lists as expandable `<details>` elements.

In [ ]:
coll = {  # => div
        'Apollo astronauts': [  # ul
            'Neil Armstrong',  # li
            'Alan Bean',  # li
            {  # li
                'other': [  # ul
                    'Bruce Wayne',  # li
                    'Clark Kent',  # li
                    'Peter Parker'  # li
                ]
            },
            'Edgar Mitchell',   # li
            'Alan Shepard'  # li
        ],
        'Apollo 11': [  # ul
            'Neil Armstrong',   # li
            'Buzz Aldrin',   # li
            'Edgar Mitchell',   # li
            {  # li
                'a':  # ul
                    1  # li
            },
            'Alan Shepard'  # li
        ]
}

In [ ]:
css = '\n'.join((await read_msg()).content.splitlines()[1:])
print(Style(css))

<style>.coll ul { list-style-type:none; list-style-position: outside; padding-inline-start: 22px; margin: 0px; }</style>


In [ ]:
#| export
def add_style(css:str):
    _stl = Div(Style(css), hx_swap_oob='beforeend:#dialog-container')
    add_html(_stl)

In [ ]:
add_style(css)

In [ ]:
%%html
<div class="coll">
<ul data-coll-name="Apollo astronauts">
  <li data-it>Neil Armstrong</li>
  <li data-it><a href="/asd">Alan Bean</a></li>
  <li data-it>other<ul data-coll-name="other">
    <li data-it>Bruce Wayne</li>
    <li data-it>Clark Kent</li>
    <li data-it>Peter Parker</li>
  </ul></li>
  <li data-it>Edgar Mitchell</li>
  <li data-it>Alan Shepard</li>
</ul>

<ul data-coll-name="Apollo 11">
  <li data-it>Neil Armstrong </li>
  <li data-it>Buzz Aldrin</li>
  <li data-it>Edgar Mitchell</li>
  <li data-it>a<ul data-coll-name="a">
    <li data-it>1</li>
  </ul></li>
  <li data-it>Alan Shepard</li>
</ul>
</div>

HTML(<div class="coll">
<ul data-coll-name="Apollo astronauts">
  <li data-it>Neil Armstrong</li>
  <li data-it><a href="/asd">Alan Bean</a></li>
  <li data-it>other<ul data-coll-name="other">
    <li data-it>Bruce Wayne</li>
    <li data-it>Clark Kent</li>
    <li data-it>Peter Parker</li>
  </ul></li>
  <li data-it>Edgar Mitchell</li>
  <li data-it>Alan Shepard</li>
</ul>

<ul data-coll-name="Apollo 11">
  <li data-it>Neil Armstrong </li>
  <li data-it>Buzz Aldrin</li>
  <li data-it>Edgar Mitchell</li>
  <li data-it>a<ul data-coll-name="a">
    <li data-it>1</li>
  </ul></li>
  <li data-it>Alan Shepard</li>
</ul>
</div>
)

In [ ]:
#| exporti
def it2lbl(it, nm=''):
    # return ((nm, it) if nm else (it,)) if isinstance(it, FT) else (str(it),)
    it = it if isinstance(it, FT) else str(it)
    return (nm, it) if nm else (it,)

In [ ]:
#| export
def list2lis(its:list, dproc, eproc=it2lbl) -> list[Li]:
    return flatten([
        [Li(data_it=True)(*eproc(ul,k)) for k,ul in dproc(it).items()] if isinstance(it, dict) else
        Li(data_it=True)(*eproc(it))
        for it in its], (FT,))

def dict2uls(d:dict, eproc=it2lbl) -> dict[Ul]:
    return {
        k:Ul(data_coll_name=k)(*list2lis(v if is_listy(v) else [v], dict2uls, eproc)) 
        for k,v in d.items()
    }

In [ ]:
uls = dict2uls(coll)
print(_ := to_xml(Div(*uls.values(), cls='coll')))
HTML(_)

In [ ]:
%%html
<div class="coll">
<details open><summary>Apollo astronauts</summary><ul data-coll-name="Apollo astronauts">
  <li data-it>Neil Armstrong</li>
  <li data-it><a href="/asd">Alan Bean</a></li>
  <li data-it><details open><summary>other</summary><ul data-coll-name="other">
    <li data-it>Bruce Wayne</li>
    <li data-it>Clark Kent</li>
    <li data-it>Peter Parker</li>
  </ul></details></li>
  <li data-it>Edgar Mitchell</li>
  <li data-it>Alan Shepard</li>
</ul></details>

<details open><summary>Apollo 11</summary><ul data-coll-name="Apollo 11">
  <li data-it>Neil Armstrong </li>
  <li data-it>Buzz Aldrin</li>
  <li data-it>Edgar Mitchell</li>
  <li data-it><details open><summary>a</summary><ul data-coll-name="a">
    <li data-it>1</li>
  </ul></details></li>
  <li data-it>Alan Shepard</li>
</ul></details>
</div>

HTML(<div class="coll">
<details open><summary>Apollo astronauts</summary><ul data-coll-name="Apollo astronauts">
  <li data-it>Neil Armstrong</li>
  <li data-it><a href="/asd">Alan Bean</a></li>
  <li data-it><details open><summary>other</summary><ul data-coll-name="other">
    <li data-it>Bruce Wayne</li>
    <li data-it>Clark Kent</li>
    <li data-it>Peter Parker</li>
  </ul></details></li>
  <li data-it>Edgar Mitchell</li>
  <li data-it>Alan Shepard</li>
</ul></details>

<details open><summary>Apollo 11</summary><ul data-coll-name="Apollo 11">
  <li data-it>Neil Armstrong </li>
  <li data-it>Buzz Aldrin</li>
  <li data-it>Edgar Mitchell</li>
  <li data-it><details open><summary>a</summary><ul data-coll-name="a">
    <li data-it>1</li>
  </ul></details></li>
  <li data-it>Alan Shepard</li>
</ul></details>
</div>
)

In [ ]:
#| exporti
def it2lbl2(it, nm=''):
    return (it,) if isinstance(it, FT) else (str(it),)

In [ ]:
#| export
def dict2dtls(d:dict, eproc=it2lbl2) -> dict[Details]:
    return {
        k:Details(open=True, data_coll_name=k)(
            Summary(k),
            Ul()(*list2lis(v if is_listy(v) else [v], dict2dtls, eproc))
        )
        for k,v in d.items()
    }

In [ ]:
uls = dict2dtls(coll)
print(_ := to_xml(Div(*uls.values(), cls='coll')))
HTML(_)

In [ ]:
def _v2tag(v):
    "Render value with appropriate CSS class based on type"
    c = 'None' if v is None else ''
    return Span(shorten(v, 'r', 140) if v is not None else 'None', cls=f"v{' '+c if c else ''}")

In [ ]:

test_eq(str(_v2tag(None)), '<span class="v None">None</span>')
test_eq(str(_v2tag(1)), '<span class="v">1</span>')

In [ ]:
_ = '\n'.join((await read_msg()).content.splitlines()[1:])
await link_msg(f"_css_coll = '''\n{_}\n'''", msg_type='code', is_exported=True);

In [ ]:
#| export
_css_coll = '''
.coll ul { list-style-type:none; list-style-position: outside; padding-inline-start: 22px; margin: 0px; }
.coll .None { color: gray; }
'''
#| linkedto: _f292d92c

In [ ]:
#| export
class DetailsColl(dict):
    "Interactive collapsible collection viewer with HTML details/summary structure"
    def __init__(self, d, summary:str='', open:bool=True, openall:bool=False):
        super().__init__(d or {})
        self.summary, self.open, self.openall = str(summary), open, openall
    def show(self): show(self)
    def __ft__(self, its=None, summary:str|None=None, lvl:int=0, open:bool=False):
        return Div(*dict2dtls(self).values(), cls='coll')
    _css_ = _css_coll
    @classmethod
    def setup_style(cls): add_style(cls._css_)

In [ ]:
# DetailsColl.setup_style()

In [ ]:
dtl = DetailsColl({'a':1})
print(to_xml(dtl))
dtl.show()

<div class="coll">
<details open data-coll-name="a"><summary>a</summary>    <ul>
      <li data-it>1</li>
    </ul>
</details></div>



HTML(<div class="coll">
<details open data-coll-name="a"><summary>a</summary>    <ul>
      <li data-it>1</li>
    </ul>
</details></div>
)

In [ ]:
dtl = DetailsColl({'a':1, 'b':2})
print(to_xml(dtl))
dtl.show()

<div class="coll">
<details open data-coll-name="a"><summary>a</summary>    <ul>
      <li data-it>1</li>
    </ul>
</details><details open data-coll-name="b"><summary>b</summary>    <ul>
      <li data-it>2</li>
    </ul>
</details></div>



HTML(<div class="coll">
<details open data-coll-name="a"><summary>a</summary>    <ul>
      <li data-it>1</li>
    </ul>
</details><details open data-coll-name="b"><summary>b</summary>    <ul>
      <li data-it>2</li>
    </ul>
</details></div>
)

In [ ]:
dtl = DetailsColl({'a':['abc', 13, {'d':['efg', 133]}, None]})
print(to_xml(dtl))
dtl.show()

<div class="coll">
<details open data-coll-name="a"><summary>a</summary>    <ul>
      <li data-it>abc</li>
      <li data-it>13</li>
      <li data-it>
<details open data-coll-name="d"><summary>d</summary>          <ul>
            <li data-it>efg</li>
            <li data-it>133</li>
          </ul>
</details>      </li>
      <li data-it>None</li>
    </ul>
</details></div>



HTML(<div class="coll">
<details open data-coll-name="a"><summary>a</summary>    <ul>
      <li data-it>abc</li>
      <li data-it>13</li>
      <li data-it>
<details open data-coll-name="d"><summary>d</summary>          <ul>
            <li data-it>efg</li>
            <li data-it>133</li>
          </ul>
</details>      </li>
      <li data-it>None</li>
    </ul>
</details></div>
)

In [ ]:
dtl = DetailsColl({
    'Apollo astronauts': [
        'Neil Armstrong',
        '<a href="/asd">Alan Bean</a>',
        'Buzz Aldrin',
        {
            'letters': [
                'a', 
                'b',
                'c', 
            ]
        },
        'Edgar Mitchell',
        'Alan Shepard'
    ]
})

test_eq(val_at(dtl, 'Apollo astronauts.5'), 'Alan Shepard')
test_eq(val_at(dtl, 'Apollo astronauts.3.letters.0'), 'a')
print(to_xml(dtl))
dtl.show()

<div class="coll">
<details open data-coll-name="Apollo astronauts"><summary>Apollo astronauts</summary>    <ul>
      <li data-it>Neil Armstrong</li>
      <li data-it>&lt;a href="/asd"&gt;Alan Bean&lt;/a&gt;</li>
      <li data-it>Buzz Aldrin</li>
      <li data-it>
<details open data-coll-name="letters"><summary>letters</summary>          <ul>
            <li data-it>a</li>
            <li data-it>b</li>
            <li data-it>c</li>
          </ul>
</details>      </li>
      <li data-it>Edgar Mitchell</li>
      <li data-it>Alan Shepard</li>
    </ul>
</details></div>



HTML(<div class="coll">
<details open data-coll-name="Apollo astronauts"><summary>Apollo astronauts</summary>    <ul>
      <li data-it>Neil Armstrong</li>
      <li data-it>&lt;a href="/asd"&gt;Alan Bean&lt;/a&gt;</li>
      <li data-it>Buzz Aldrin</li>
      <li data-it>
<details open data-coll-name="letters"><summary>letters</summary>          <ul>
            <li data-it>a</li>
            <li data-it>b</li>
            <li data-it>c</li>
          </ul>
</details>      </li>
      <li data-it>Edgar Mitchell</li>
      <li data-it>Alan Shepard</li>
    </ul>
</details></div>
)

In [ ]:
tr = list(dir_tree('.'))
nbs_dtl = dtl = DetailsColl({'nbs': tr})
nbs_dtl.show()

HTML(<div class="coll">
<details open data-coll-name="nbs"><summary>nbs</summary>    <ul>
      <li data-it>00_basic.ipynb</li>
      <li data-it>00_dutil.ipynb</li>
      <li data-it>00_project.ipynb</li>
      <li data-it>01_display.ipynb</li>
      <li data-it>01_display_corrupted.ipynb</li>
      <li data-it>01_flakes.ipynb</li>
      <li data-it>02_git.ipynb</li>
      <li data-it>02_isolated.ipynb</li>
      <li data-it>02_logger.ipynb</li>
      <li data-it>02_logger_loguru.ipynb</li>
      <li data-it>02_server.ipynb</li>
      <li data-it>05_test.ipynb</li>
      <li data-it>10_callback.ipynb</li>
      <li data-it>15_config.ipynb</li>
      <li data-it>17_display.ipynb</li>
      <li data-it>20_widgets.ipynb</li>
      <li data-it>_quarto.yml</li>
      <li data-it>apilist.txt</li>
      <li data-it>
<details open data-coll-name="data"><summary>data</summary>          <ul>
            <li data-it>data/log_across.ipynb</li>
            <li data-it>data/test.ipynb</li>
          </ul>
</details>      </li>
      <li data-it>index.ipynb</li>
      <li data-it>llms-ctx-full.txt</li>
      <li data-it>llms-ctx.txt</li>
      <li data-it>llms.txt</li>
      <li data-it>nbdev.yml</li>
      <li data-it>
<details open data-coll-name="static"><summary>static</summary>          <ul>
            <li data-it>static/apollo_astronauts.json</li>
            <li data-it>static/file.txt</li>
            <li data-it>static/wordlist.txt</li>
          </ul>
</details>      </li>
      <li data-it>styles.css</li>
    </ul>
</details></div>
)

## DetailsJSON
> Collapsible JSON/dict viewer component for notebooks

A FastHTML component that renders dictionaries and nested data as expandable `<details>` elements with syntax highlighting.

In [ ]:
%%HTML
<style>
    details.json ul { list-style-type:none; list-style-position: outside; padding-inline-start: 22px; margin: 0px; }
</style>
<details open class="json">
<summary>Apollo astronauts</summary>
<ul>
  <li><span>1</span>: Neil Armstrong</li>
  <li><span>2</span>: Alan Bean</li>
  <li><details>
<summary>Apollo 11</summary>
<ul>
  <li><span>1</span>: Neil Armstrong</li>
  <li><span>2</span>: Alan Bean</li>
  <li><div><span>3</span>: Buzz Aldrin</div></li>
  <li><span>4</span>: Edgar Mitchell</li>
  <li><span>5</span>: Alan Shepard</li>
</ul></li>
  <li><span>4</span>: Edgar Mitchell</li>
  <li><span>5</span>: Alan Shepard</li>
</ul>

</details>

In [ ]:
#| exporti
def Val(v): 
    "Render value with appropriate CSS class based on type"
    c = (
        'null' if v is None else 
        'true' if v is True else 
        'false' if v is False else 
        'string' if isinstance(v, str) else 
        'number' if isinstance(v, (int, float)) else 
        '')
    return Span(shorten(v, 'r', 140) if v is not None else 'None', cls=f"v {c}")

In [ ]:
#| exporti
def NameVal(k, v):
    "Render key-value pair with name and value styling"
    return Span(Span(k, cls='n'), ': ', Val(v))

In [ ]:
#| export
_css_json = (
    'details.json ul { list-style-type:none; list-style-position: outside; padding-inline-start: 22px; margin: 0; } '
    '''details.json .string { color: #24837b; } details.json .string::before { content: "'"; } details.json .string::after { content: "'"; } '''
    'details.json .number { color: #ad8301; } '
    'details.json .true { color: blue; } '
    'details.json .false { color: red; } '
    'details.json .null { color: gray; } '
    'details.json span.n { color: darkgrey; } '
)


In [ ]:
#| export
class DetailsJSON(dict):
    "Interactive collapsible JSON viewer with HTML details/summary structure"
    def __init__(self, *args, summary:str='', open:bool=True, openall:bool=False, skip:Sequence[str]=(), **kwargs):
        super().__init__(*args, **kwargs)
        self.summary, self.open, self.openall, self.skip = str(summary), open, openall, skip
    def show(self): show(self)
    def __ft__(self, d:Mapping|None=None, summary:str|None=None, lvl:int=0, open:bool=False):
        if d is None: d = self; summary = self.summary or 'summary'; open=self.open
        open = self.openall or open
        return (
            Details(open=open, cls="json")(
                Summary(summary, _n),
                Ul()(*(
                    Li(NameVal(k, v)) if k in self.skip else
                    self.__ft__(v, k, lvl+1) if isinstance(v, Mapping) else
                    self.__ft__(dict(list(zip(range(len(v)), v))), k, lvl+1) if is_listy(v) else
                    Li(NameVal(k, v)) 
                    for k,v in d.items()))))
    _css_ = _css_json
    @classmethod
    def setup_style(cls): add_style(cls._css_)

In [ ]:
#| export
DetailsJSON.setup_style()

In [ ]:
dtl = DetailsJSON({
    '1': 'Neil Armstrong',
    '2': 'Alan Bean',
    '3': 'Buzz Aldrin',
    'letters': {
        'a': 1, 
        'b': 2
        },
    '5': 'Edgar Mitchell',
    '6': 'Alan Shepard'
}, summary='Apollo astronauts')

test_eq(val_at(dtl, '5'), 'Edgar Mitchell')
test_eq(val_at(dtl, 'letters.a'), 1)
print(to_xml(dtl))
dtl.show()

In [ ]:
d= {
    "idx": 1,
    "cell_type": "code",
    "source": "# cell 1\nprint('hello')",
    "id": "W1sZmlsZQ==",
    "metadata": {
        "brd": {
            "id": "717322f8-95fa-425c-839d-8b9e7d4ef921"
        }
    },
    "outputs": [
        {'output_type': 'stream', 'name': 'stdout', 'text': '1\n'},
        {'output_type': 'stream', 'name': 'stdout', 'text': '2\n'}
    ],
    "execution_count": 1
}
DetailsJSON(d, openall=True).show()

In [ ]:
request = {
    'headers': {
        'HX-Request': 'true',
        'HX-Current-URL': 'vscode-webview://1ql27...enderer'
    },
    'headerNames': {
        'hx-request': 'HX-Request',
        'hx-current-url': 'HX-Current-URL'
    },
    'status': 0,
    'method': 'GET',
    'url': '/DetailsJSON_5096628128/Apollo 11/Buzz Aldrin',
    'async': True,
    'timeout': 0,
    'withCredentials': False,
    'body': None,
    'req_id': '68ffadb0-958d-4346-b314-d9d62ca247d7'
}

response = {
    'headers': {
        'content-length': '441',
        'content-type': 'text/html; charset=utf-8',
        'last-modified': 'Fri, 15 Nov 2024 16:22:15 GMT',
        'cache-control': 'no-store, no-cache, must-revalidate'
    },
    'status': 200,
    'statusText': 'OK',
    'data': '<details open><summary>Buzz Aldrin</summary>\n   <ul>\n     '
'<li>Pilot on Gemini 12 and Lunar Module pilot on Apollo 11.</li>\n     '
'<li>Aldrin was the second person to walk on the moon.</li>\n     <li>The maiden '
'name of Aldrin&#x27;s mother was &quot;Moon.&quot;</li>\n     <li>While Neil was '
'the first human to step onto the moon, I&#x27;m the first alien from another '
'world to enter a spacecraft that was going to Earth.</li>\n   '
'</ul>\n</details>',
    'xml': None,
    'finalUrl': 'http://nb/DetailsJSON_5096628128/Apollo%2011/Buzz%20Aldrin',
    'req_id': '68ffadb0-958d-4346-b314-d9d62ca247d7'
}

In [ ]:
req = DetailsJSON(request, summary='request', open='all')
req.show()

In [ ]:
resp = DetailsJSON(response, summary='response')
resp.show()

# export -

In [ ]:
from pote.flakes import show_flakes
await show_flakes()

<div class="prose">

No warnings to report

</div>

In [ ]:
# #|hide
# #|eval: false
# from pote.dutil import dlg_export
# dlg_export()